### PR TITLE
API 7.0

### PR DESCRIPTION
**Pull request by MKRhere**
This is a massive API update. This was independently updated from the [Bot API changelog](https://core.telegram.org/bots/api-changelog#december-29-2023), and cross-verified against https://github.com/grammyjs/types/pull/32.

### CODE REVIEW

**Review by hypernova7 — APPROVED** [review]

File: message.ts [hunk 1]
```diff
@@ -478,6 +520,222 @@ export type MessageEntity =
...
+  | ExternalReplyInvoice
+  | ExternalReplyLocation
+  | ExternalReplyPoll
+  | ExternalReplyVenue;
+
+/** Describes reply parameters for the message that is being sent. */
+export interface ReplyParameters {
+  /** Identifier of the message that will be replied to in the current chat, or in the chat chat_id if it is specified */
+  message_id: number;
+  /** If the message to be replied to is from a different chat, unique identifier for the chat or username of the channel (in the format @channelusername) */
```
Comment (dcdunkan) [inline message.ts]:
```suggestion
  /** If the message to be replied to is from a different chat, unique identifier for the chat or username of the channel (in the format `@channelusername`) */
```

File: message.ts [hunk 2]
```diff
@@ -478,6 +520,222 @@ export type MessageEntity =
   | MessageEntity.TextLinkMessageEntity
   | MessageEntity.TextMentionMessageEntity;
 
+/** This object contains information about the quoted part of a message that is replied to by the given message. */
+export interface TextQuote {
+  /** Text of the quoted part of a message that is replied to by the given message */
+  text: string;
+  /** Special entities that appear in the quote. Currently, only bold, italic, underline, strikethrough, spoiler, and custom_emoji entities are kept in quotes. */
+  entities: MessageEntity[];
```
Comment (dcdunkan) [inline message.ts]:
```suggestion
  entities?: MessageEntity[];
```
`entities` is also marked as optional: https://core.telegram.org/bots/api#textquote

File: methods.ts [hunk 3]
```diff
@@ -472,8 +492,8 @@ export type ApiMethods<F> = {
     disable_notification?: boolean;
     /** Protects the contents of the sent messages from forwarding and saving */
     protect_content?: boolean;
-    /** If messages are a reply, ID of the original message */
-    reply_to_message_id?: number;
+    /** Description of the message to reply to */
+    reply_parameters?: ReplyParameters;
     /** Pass True if the message should be sent even if the specified replied-to message is not found */
     allow_sending_without_reply?: boolean;
```
Comment (dcdunkan) [inline methods.ts]:
```suggestion
```

This was also replaced by `reply_parameters`.

File: methods.ts [hunk 4]
```diff
@@ -204,9 +220,27 @@ export type ApiMethods<F> = {
       | ForceReply;
   }): MessageId;
 
+  /** Use this method to copy messages of any kind. If some of the specified messages can't be found or copied, they are skipped. Service messages, giveaway messages, giveaway winners messages, and invoice messages can't be copied. A quiz poll can be copied only if the value of the field correct_option_id is known to the bot. The method is analogous to the method forwardMessages, but the copied messages don't have a link to the original message. Album grouping is kept for copied messages. On success, an array of MessageId of the sent messages is returned. */
+  copyMessages(args: {
+    /** Unique identifier for the target chat or username of the target channel (in the format @channelusername) */
```
Comment (dcdunkan) [inline methods.ts]:
```suggestion
    /** Unique identifier for the target chat or username of the target channel (in the format `@channelusername`) */
```

File: methods.ts [hunk 5]
```diff
@@ -204,9 +220,27 @@ export type ApiMethods<F> = {
...
+    chat_id: number | string;
+    /** Unique identifier for the target message thread (topic) of the forum; for forum supergroups only */
+    message_thread_id?: number;
+    /** Unique identifier for the chat where the original messages were sent (or channel username in the format @channelusername) */
```
Comment (dcdunkan) [inline methods.ts]:
```suggestion
    /** Unique identifier for the chat where the original messages were sent (or channel username in the format `@channelusername`) */
```

File: message.ts [hunk 6]
```diff
@@ -815,6 +1073,79 @@ export interface VideoChatParticipantsInvited {
...
+  /** True, if only users who join the chats after the giveaway started should be eligible to win */
+  only_new_members?: true;
+  /** True, if the list of giveaway winners will be visible to everyone */
+  has_public_winners?: true;
+  /** Description of additional giveaway prize */
+  prize_description?: string;
+  /** A list of two-letter ISO 3166-1 alpha-2 country codes indicating the countries from which eligible users for the giveaway must come. If empty, then all users can participate in the giveaway. Users with a phone number that was bought on Fragment can always participate in giveaways. */
+  country_codes?: string[];
+  /** Optional. The number of months the Telegram Premium subscription won from the giveaway will be active for */
+  premium_subscription_month_count: number;
```
Comment (dcdunkan) [inline message.ts]:
```suggestion
  /** The number of months the Telegram Premium subscription won from the giveaway will be active for */
  premium_subscription_month_count?: number;
```

Optional. https://core.telegram.org/bots/api#giveaway

File: message.ts [hunk 7]
```diff
@@ -815,6 +1073,79 @@ export interface VideoChatParticipantsInvited {
...
+  /** Total number of winners in the giveaway */
+  winner_count: number;
+  /** List of up to 100 winners of the giveaway */
+  winners: User[];
+  /** Optional. The number of other chats the user had to join in order to be eligible for the giveaway */
+  additional_chat_count: number;
+  /** Optional. The number of months the Telegram Premium subscription won from the giveaway will be active for */
+  premium_subscription_month_count: number;
+  /** Optional. Number of undistributed prizes */
+  unclaimed_prize_count: number;
```
Comment (dcdunkan) [inline message.ts]:
```suggestion
  /** The number of other chats the user had to join in order to be eligible for the giveaway */
  additional_chat_count?: number;
  /** The number of months the Telegram Premium subscription won from the giveaway will be active for */
  premium_subscription_month_count?: number;
  /** Number of undistributed prizes */
  unclaimed_prize_count?: number;
```

File: message.ts [hunk 8]
```diff
@@ -815,6 +1073,79 @@ export interface VideoChatParticipantsInvited {
...
+  /** Description of additional giveaway prize */
+  prize_description?: string;
+}
+
+/** This object represents a service message about the completion of a giveaway without public winners. */
+export interface GiveawayCompleted {
+  /** Number of winners in the giveaway */
+  winner_count: number;
+  /** Optional. Number of undistributed prizes */
+  unclaimed_prize_count: number;
```
Comment (dcdunkan) [inline message.ts]:
```suggestion
  /** Number of undistributed prizes */
  unclaimed_prize_count?: number;
```

File: methods.ts [hunk 9]
```diff
@@ -1146,6 +1168,14 @@ export type ApiMethods<F> = {
     language_code?: string;
   }): BotName;
 
+  /** Use this method to get the list of boosts added to a chat by a user. Requires administrator rights in the chat. Returns a UserChatBoosts object. */
+  getUserChatBoosts(arg: {
+    /** Unique identifier for the chat or username of the channel (in the format @channelusername) */
```
Comment (dcdunkan) [inline methods.ts]:
```suggestion
    /** Unique identifier for the chat or username of the channel (in the format `@channelusername`) */
```

**Review by dcdunkan — CHANGES_REQUESTED** [review]

**Pull request merged**